compiler/ballerina-parser/src/main/java/io/ballerina/compiler/internal/parser/DocumentationParser.java

```diff
@@ -157,13 +158,124 @@ private STNode parseDocumentationReference() {
         }
 
         STNode startBacktick = parseBacktickToken();
-        STNode backtickContent = parseBacktickContent();
+        STNode backtickContent = parseBacktickContent(referenceType);
         STNode endBacktick = parseBacktickToken();
-
         return STNodeFactory.createDocumentationReferenceNode(referenceType, startBacktick, backtickContent,
                 endBacktick);
     }
 
+    /**
+     * Represents the current position with respect to the head in a token-sequence-search.
+     */
+    private static class Lookahead {
+        private int offset = 1;
+    }
+
+    /**
+     * Genre of the reference that precedes the backtick block.
+     */
+    private enum ReferenceGenre {
+        NO_KEY, SPECIAL_KEY, FUNCTION_KEY
+    }
+
+    /**
+     * Look ahead and see if upcoming token sequence is valid.
+     *
+     * @param refGenre Genre of the backtick block reference
+     * @return <code>true</code> if content is valid<code>false</code> otherwise.
+     */
+    private boolean isValidBacktickContentSequence(ReferenceGenre refGenre) {
+        boolean hasMatch;
+        Lookahead lookahead = new Lookahead();
+        switch (refGenre) {
+            case SPECIAL_KEY:
+                // Look for x, m:x match
+                hasMatch = hasQualifiedIdentifier(lookahead);
+                break;
+            case FUNCTION_KEY:
+                // Look for x, m:x, x(), m:x(), T.y(), m:T.y() match
+                hasMatch = hasBacktickExpr(lookahead, true);
+                break;
+            case NO_KEY:
+                // Look for x(), m:x(), T.y(), m:T.y() match
+                hasMatch = hasBacktickExpr(lookahead, false);
+                break;
+            default:
+                throw new IllegalStateException("Unsupported backtick reference genre");
+        }
+
+        return hasMatch && peek(lookahead.offset).kind == SyntaxKind.BACKTICK_TOKEN;
```

Comment:
If we can return `false` from the calling methods ie `hasQualifiedIdentifier` and `hasBacktickExpr` whenever `peek(lookahead.offset).kind == SyntaxKind.BACKTIC_CONTENT` then we can simplify this by removing the `Lookahead` class

Reply:
As we discussed offline problem is, methods like hasIdentifier is called in other places as well. eg. hasQualifiedIdentifer, hasBacktickExpr.